conda/core/path_actions.py

```diff
@@ -82,27 +89,84 @@
 )
 
 
-class _Action(metaclass=ABCMeta):
+class Action:
+    """Base class for path manipulation actions, including linking, unlinking, and others.
+
+    Pre and post-transaction plugins should inherit this class to implement their
+    own verification, execution, reversing, and cleanup steps. These methods are
+    guaranteed to be called in the following order:
+
+        1. ``verify``
+        2. ``execute``
+        3. ``reverse`` (only if ``execute`` raises an exception)
+        4. ``cleanup``
+
+
+    :param transaction_context: Mapping between target prefixes and PrefixActionGroupV2
+        instances
+    :param target_prefix: Target prefix for the action
+    :param unlink_precs: Package records to be unlinked
+    :param link_precs: Package records to link
+    :param remove_specs: Specs to be removed
+    :param update_specs: Specs to be updated
+    :param neutered_specs: Specs to be neutered
+    """
+
     _verified = False
 
+    def __init__(
+        self,
+        transaction_context: dict[str, str] | None = None,
+        target_prefix: str | None = None,
+        unlink_precs: Iterable[PackageRecord] | None = None,
+        link_precs: Iterable[PackageRecord] | None = None,
+        remove_specs: Iterable[MatchSpec] | None = None,
+        update_specs: Iterable[MatchSpec] | None = None,
+        neutered_specs: Iterable[MatchSpec] | None = None,
+    ):
+        self.transaction_context = transaction_context
+        self.target_prefix = target_prefix
+        self.unlink_precs = unlink_precs
+        self.link_precs = link_precs
+        self.remove_specs = remove_specs
+        self.update_specs = update_specs
+        self.neutered_specs = neutered_specs
+
     @abstractmethod
-    def verify(self):
-        # if verify fails, it should return an exception object rather than raise
-        #  at the end of a verification run, all errors will be raised as a CondaMultiError
-        # after successful verification, the verify method should set self._verified = True
-        raise NotImplementedError()
+    def verify(self) -> Exception | None:
+        """Carry out any pre-execution verification.
+
+        Should set self._verified = True upon success.
+
+        :return: On failure, this function should return (not raise!) an exception
+        object. At the end of the verification run, all errors will be raised as a
+        CondaMultiError.
+        """
 
     @abstractmethod
-    def execute(self):
-        raise NotImplementedError()
+    def execute(self) -> None:
+        """Execute the action.
+
+        Called after ``self.verify()``. If this function raises an exception,
+        ``self.reverse()`` will be called.
+        """
+
+    def __call__(self) -> None:
+        """Execute the action."""
+        return self.execute()
```

Comment:
I don't think this serves a purpose anymore, correct? If so we should revert this and just call `action.execute` as before.